File: DaNang_2324/Bai4/TANGQUA.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

int longest_increasing_subsequence(const vector<int>& sequence) {
    vector<int> lis;
    for (int value : sequence) {
        auto pos = lower_bound(lis.begin(), lis.end(), value);
        if (pos == lis.end()) {
            lis.push_back(value);
        } else {
            *pos = value;
        }
    }
    return lis.size();
}

int main() {
    ifstream infile("TANGQUA.INP");
    ofstream outfile("TANGQUA.OUT");

    int n;
    infile >> n;
    vector<int> values(n);

    for (int i = 0; i < n; i++) {
        infile >> values[i];
    }

    int lis_length = longest_increasing_subsequence(values);
    int result = n - lis_length;

    outfile << result << endl;

    infile.close();
    outfile.close();

    return 0;
}

File: THCS/2023_2024/DaNang_2324/Bai4/TANGQUA.py
input_path = "./TANGQUA.INP"
output_path = "./TANGQUA.OUT"
def longest_increasing_subsequence(sequence):
    from bisect import bisect_left

    lis = []
    for value in sequence:
        pos = bisect_left(lis, value)
        if pos == len(lis):
            lis.append(value)
        else:
            lis[pos] = value
    return len(lis)

def main():
    with open(input_path, "r") as infile:
        n = int(infile.readline().strip())
        values = list(map(int, infile.readline().strip().split()))
    
    lis_length = longest_increasing_subsequence(values)
    result = n - lis_length
    
    with open(output_path, "w") as outfile:
        outfile.write(str(result) + "\n")

if __name__ == "__main__":
    main()
